zookeeper/core/partial_component.py

```diff
@@ -0,0 +1,112 @@
+import inspect
+from typing import Generic, Type, TypeVar
+
+from zookeeper.core import utils
+
+# A type-variable used to parameterise `PartialComponent`. This is for static
+# type-checking.
+_ComponentType = TypeVar("_ComponentType")
+
+
+# Defined once here to avoid having to define it twice in the `__init__` body.
+_kwargs_error = TypeError(
+    "Keyword arguments passed to `PartialComponent` must be either:\n"
+    "- An immutable value (int, float, bool, string, or None).\n"
+    "- A function or lambda accepting no arguments and returning the \n"
+    "  value that should be passed to the component upon instantiation.\n"
+    "- Another `PartialComponent`."
+    "Wrapping non-immutable values in a function / lambda allows the values "
+    "to be lazily evaluated; they won't be created at all if the partial "
+    "component is never instantiated."
+)
+
+
+class PartialComponent(Generic[_ComponentType]):
```

Comment:
This is pretty cool, I guess we can get rid of the LayerFactories in object detection now